pkg/cmd/admin/image/verify-signature.go

```diff
@@ -0,0 +1,259 @@
+	if len(o.PublicKeyFilename) > 0 {
+		if o.Remove {
+			return kcmdutil.UsageError(cmd, "cannot use public key when removing verification status")
+		}
+		if o.PublicKey, err = ioutil.ReadFile(o.PublicKeyFilename); err != nil {
+			return err
+		}
+	}
+	if o.Client, _, err = f.Clients(); err != nil {
+		return err
+	}
+	// Only make this API call when we are sure we will be writing validation.
+	if o.Confirm && !o.Remove {
+		if me, err := o.Client.Users().Get("~"); err != nil {
+			return err
+		} else {
+			o.CurrentUser = me.Name
+		}
+	}
+
+	return nil
+}
+
+// verifySignature verifies the image signature and returns the identity when the signature
+// is valid.
+// TODO: This should be calling the 'containers/image' library in future.
+func (o *VerifyImageSignatureOptions) verifySignature(signature []byte) (string, []byte, error) {
+	var (
+		mechanism SigningMechanism
+		err       error
+	)
+	// If public key is specified, use JUST that key for verification. Otherwise use all
+	// keys in local GPG public keyring.
+	if len(o.PublicKeyFilename) == 0 {
+		mechanism, err = newGPGSigningMechanismInDirectory("")
+	} else {
+		mechanism, _, err = newEphemeralGPGSigningMechanism(o.PublicKey)
+	}
+	if err != nil {
+		return "", nil, err
+	}
+	defer mechanism.Close()
+	content, identity, err := mechanism.Verify(signature)
+	if err != nil {
+		return "", nil, err
+	}
+	return string(identity), content, nil
+}
+
+// verifySignatureContent verifies that the signature content matches the given image.
+// TODO: This should be done by calling the 'containers/image' library in future.
+func (o *VerifyImageSignatureOptions) verifySignatureContent(content []byte) (string, string, error) {
+	// TODO: The types here are just to decompose the JSON. The fields should not change but
+	// we need to use containers/image library here to guarantee compatibility in future.
+	type criticalImage struct {
+		Digest string `json:"docker-manifest-digest"`
+	}
+	type criticalIdentity struct {
+		DockerReference string `json:"docker-reference"`
+	}
+	type critical struct {
+		Image    criticalImage    `json:"image"`
+		Identity criticalIdentity `json:"identity"`
+	}
+	type message struct {
+		Critical critical `json:"critical"`
+	}
+	m := message{}
+	if err := json.Unmarshal(content, &m); err != nil {
+		return "", "", err
+	}
+	if o.InputImage != m.Critical.Image.Digest {
+		return "", "", fmt.Errorf("signature is valid for digest %q not for %q", m.Critical.Image.Digest, o.InputImage)
+	}
+	return m.Critical.Image.Digest, m.Critical.Identity.DockerReference, nil
+}
+
+// verifyImageIdentity verifies the source of the image specified in the signature is
+// valid.
+func (o *VerifyImageSignatureOptions) verifyImageIdentity(reference, digest string) error {
+	ref, err := imageapi.ParseDockerImageReference(reference)
+	if err != nil {
+		return err
+	}
+	// Verify the tag exists
+	tag, err := o.Client.ImageStreamTags(ref.Namespace).Get(ref.Name, ref.Tag)
```

Comment:
References either with a tag or a digest can happen. `NameOnly` references should not.

(In principle, an attacker can put anything invalid in there, including empty strings, completely invalid reference formats, `NameOnly` references, or references which contain both a tag and a digest.)